SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -0,0 +1,338 @@
+        [YapDatabaseViewGrouping withObjectBlock:^NSString *_Nullable(YapDatabaseReadTransaction *_Nonnull transaction,
+            NSString *_Nonnull collection,
+            NSString *_Nonnull key,
+            id _Nonnull object) {
+            if (![object isKindOfClass:[OWSMessageProcessingJob class]]) {
+                OWSFail(@"Unexpected object: %@ in collection: %@", object, collection) return nil;
+            }
+
+            // Arbitrary string - all in the same group. We're only using the view for sorting.
+            return OWSMessageProcessingJobFinderExtensionGroup;
+        }];
+
+    YapDatabaseViewOptions *options = [YapDatabaseViewOptions new];
+    options.allowedCollections =
+        [[YapWhitelistBlacklist alloc] initWithWhitelist:[NSSet setWithObject:[OWSMessageProcessingJob collection]]];
+
+    return [[YapDatabaseView alloc] initWithGrouping:grouping sorting:sorting versionTag:@"1" options:options];
+}
+
+
++ (void)syncRegisterDatabaseExtension:(YapDatabase *)database
+{
+    YapDatabaseView *existingView = [database registeredExtension:OWSMessageProcessingJobFinderExtensionName];
+    if (existingView) {
+        // already initialized
+        return;
+    }
+    [database registerExtension:[self databaseExension] withName:OWSMessageProcessingJobFinderExtensionName];
+}
+
+@end
+
+#pragma mark - Queue Processing
+
+@interface OWSMessageProcessingQueue : NSObject
+
+@property (nonatomic, readonly) TSMessagesManager *messagesManager;
+@property (nonatomic, readonly) OWSMessageProcessingJobFinder *finder;
+
+- (instancetype)initWithMessagesManager:(TSMessagesManager *)messagesManager
+                                 finder:(OWSMessageProcessingJobFinder *)finder NS_DESIGNATED_INITIALIZER;
+- (instancetype)init NS_UNAVAILABLE;
+
+@end
+
+@implementation OWSMessageProcessingQueue
+
+- (instancetype)initWithMessagesManager:(TSMessagesManager *)messagesManager
+                                 finder:(OWSMessageProcessingJobFinder *)finder
+{
+    OWSSingletonAssert();
+
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _messagesManager = messagesManager;
+    _finder = finder;
+
+    return self;
+}
+
+#pragma mark - instance methods
+
+- (void)enqueueEnvelopeForProcessing:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    [self.finder addJobForEnvelope:envelope];
+}
+
+- (void)drainQueue
+{
+    dispatch_async(self.class.serialGCDQueue, ^{
+        OWSMessageProcessingJob *_Nullable job = [self.finder nextJob];
+        if (job == nil) {
+            DDLogVerbose(@"%@ Queue is drained", self.tag);
+            return;
+        }
+
+        [self processJob:job
+              completion:^{
+                  [self drainQueue];
+              }];
+    });
+}
+
+- (void)processJob:(OWSMessageProcessingJob *)job completion:(void (^)())completion
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self.messagesManager processEnvelope:job.envelopeProto
```

Comment:
In the case of "unknown group" messages, I think that's problematic. We can't re-decrypt a message. Once we decrypt it, we throw away the key, so we'd have to persist the decrypted portion.

Maybe we ultimately need to split this into a DecryptEnvelopeJob and a ProcessDecryptedEnvelopeJob (names are a WIP 😉 ). Doing that would require a non-trivial refactor of TSMessagesManager, but I think it would be appropriate. We have way too much Signal-iOS specific app logic in TSMessagesManager as is.

Does that seem like the right approach?

And if so, does merging this as is move us closer, farther, laterally from that goal?

I feel like it moves us a little closer, in that now we have a persistent queue of encrypted envelopes. The next step would be splitting out some functionality and creating a persistent store of decrypted envelopes (waiting to be processed into something user visible).